### PR TITLE
Typeform Data Source — Migration

### PR DESCRIPTION
**Pull request by Offir**
[Asana Task — TypeForm data source: 404 Error on load](https://app.asana.com/0/280843276935180/797693123667337/f)
[Asana Task — Typeform data source error handling](https://app.asana.com/0/280843276935180/668459744820582/f)

## Description
Migrate from Data API to Responses API
https://developer.typeform.com/responses/

#### Pagination
- Fixed the pagination to work as it should to:
  * Changed from `after` to `before`.
  * Added a new `sort` parameter.
  * Added an optional `completed`.

#### Additional
- Changed `pep8` to `pycodestyle`.
- Changed the version to `1.1.0`.
- Upgraded the version of modules:
  * `requests`
  * `backoff`
  * `coverage`
- Improved coverage and fixed tests.
- Changed `SECOND` to `LIMIT_PERIOD_SEC`.
- Improved `read()` functionality:
  * It should not return empty `[]`.
  * Instead, move on to other forms.
  * Only at the end return `None`.
- Made some cosmetic changes.

**Note:** Please remember to tag the `1.1.0` version after merging and change the `sources.txt` file.

## Checklist
- [ ] Added to CHANGELOG?
- [ ] Staging tests?
- [ ] Migration script?

### CODE REVIEW

File: typeform/__init__.py
Comment (Offir) [inline typeform/__init__.py:12]: Typeform uses OAuth 2.0

File: .travis.yml
Comment (yonivy) [inline .travis.yml]: If we're here already let's update pep8 to pycodestyle

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py]: please change to `source.get('destination', DESTINATION)` to keep consistent with other sources

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py:94]: why the switch from name to title?

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py]: SECOND may never change its value, which is kinda counter productive. Name it something like `LIMIT_PERIOD_SEC = 1` and then you could easily change the value.

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py:186]: I've seen this code before... we may want to add it to the base class itself.

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py:161]: this seems different than the old id

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py:85]: I don't see 'total' in the updated code. Did we plan to remove it?

File: typeform/typeform.py
Comment (Offir) [reply typeform/typeform.py:94]: the response from `https://api.typeform.com/forms`:
```{
    "total_items": 2,
    "page_count": 1,
    "items": [
        {
            "id": "bYABaN",
            "title": "Party Invitation (copy)",
             ... more attributes (but no `name` attribute)
        },
        {
            "id": "dF9Wrj",
            "title": "Ofir sign",
           ... more attributes (but no `name` attribute)
    ]
}
```

As you can see there aren't `name` attributes

File: typeform/typeform.py
Comment (Offir) [reply typeform/typeform.py:85]: Theres isn't `total` attribute anymore.
There is `total_items` attribute (is the number of responses your request retrieved)
When we paginate `total_items` is changing on each response.
I return only the items data from the `read()` method
The response looks like this:
```
{
    "total_items": XX,
    "page_count": XX,
    "items": [
        ....
     ]
}
```
https://developer.typeform.com/responses/JSON-response-explanation/

File: typeform/typeform.py
Comment (Offir) [reply typeform/typeform.py:161]: This is the id of an inner attribute. For example:
```
"field": {
   "id": "tPeFPGJBIeI1",
   "type": "short_text",
   "ref": "5e0e49852060b732"
 },
```
Every form response (that is completed) have the same question so we need to make it unique.  

The `id` of every item is the `token` attribute (or `landing_id` attrbiute - they are the same)

File: .travis.yml
Comment (Offir) [reply .travis.yml]: I tried to use `pycodestyle` but  the build fails (see previous commit).
I don't wont to spend more time on it and to try to solve the issue so i changed it back to pep8.

**Review by alonbrody — COMMENTED** [review]
LGTM

However I do know that we are waiting for a response from TypeForm so I'm waiting with the official approval

File: .travis.yml
Comment (german-panoply) [reply .travis.yml]: Changed to `pycodestyle`.

File: typeform/typeform.py
Comment (german-panoply) [reply typeform/typeform.py]: Changed.

File: typeform/typeform.py
Comment (german-panoply) [reply typeform/typeform.py]: Changed.

File: typeform/typeform.py
Comment (yonivy) [inline typeform/typeform.py:119]: Include both what?

**Review by yonivy — APPROVED** [review]
Based on my previous review and seeing that my comments were addressed I'm approving this. But, I'm a bit confused regarding the changes made from the last time this was reviewed so just to make sure @Offir or @german-panoply please answer me this:

1. Do we have an approval from Typeform that the pagination issue is resolved?
2. Can you briefly explain how the new mechanism works? I see they use `before` which sounds rather strange to our use case.

**Comment by Offir** [comment]
@yonivy  Typeform support offer us a workaround. 
This is the email I got from them:
"
In this case, the first response with id '4c90b0c0b7de1dca9d44779c85fff8fa' is not the first persisted, that is why the next one is a landing. 

In this case, to make it easier for the customer, I would suggest changing the first request to https://api.typeform.com/forms/bYABaN/responses?page_size=1&completed=1 and thenhttps://api.typeform.com/forms/bYABaN/responses?page_size=1&&completed=1&before=4c90b0c0b7de1dca9d44779c85fff8fa since the default order is from most recent submissions to the oldest. 

Meaning that 4c90b0c0b7de1dca9d44779c85fff8fa is actually the latest one to get into the system, not the first (hence before makes more sense). 

In case the customer is interested also in getting landings, I would change the first request to https://api.typeform.com/forms/bYABaN/responses?page_size=1&sort=landed_at,desc and then he can paginate with https://api.typeform.com/forms/bYABaN/responses?page_size=1&&completed=1&before=<token>
"

File: typeform/typeform.py
Comment (german-panoply) [reply typeform/typeform.py:119]: - `'completed': 1` returns only completed responses, which have a list of `answers`.
- `'completed': 0` returns only not completed responses with `answers` as `null`.
- Not specifying  the `completed` parameter at all  returns both types of responses.

**Comment by german-panoply** [comment]
@yonivy, in short, Typeform support talked to their developers and confirmed that it's the way the pagination should work. 

> Hi again Ofir and thanks for your patience while we looked into this. I needed to check with our Devs on this one so I'm sorry for the delay!
> 
> They have let me know that this is, in fact, expected behaviour and, as per our documentation, after/before does not necessarily match submission dates, but instead follow a time-stamp that has been stored. 
>
> They did have the following workaround, though, which I hope helps:
> [...]

#### Explanation:
- `before` and `after` based on time-stamps, which means in order for pagination to work properly, we must use the `sort` parameter to sort the data by `landed_at` in `desc`ending or `asc`ending mode.
- Some default sorting options might be applied in case we include the `completed` parameter, and none are used if neither `completed` nor `sort` is included. The `sort` parameter gives us more control.

#### Example:
When using `'sort': 'landed_at,desc'`

- **Page 1 - First Request**
  * Response 23
  * Response 22
  * Response 21
  * Response 20
  * Response 19

Then when adding `'before': '19'`

- **Page 2 - Second Request**
  * Response 18
  * Response 17
  * Response 16
  * Response 15
  * Response 14

**Pull request merged**
